spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/StreamListenerAnnotationBeanPostProcessor.java

```diff
@@ -109,84 +120,138 @@ public Object postProcessAfterInitialization(final Object bean, String beanName)
 					}
 					else {
-						for (int i = 0; i < parameterTypes.length; i++) {
-							MethodParameter methodParameter = MethodParameter.forMethodOrConstructor(method, i);
-							Assert.isTrue(methodParameter.getParameterAnnotation(Input.class) != null ^
-											methodParameter.getParameterAnnotation(Output.class) != null,
-									"A declarative @StreamListener method must have its parameters annotated" +
-											"with @Input or @Output, but not with both.");
-						}
-						if (!method.getReturnType().equals(Void.TYPE)) {
-							Assert.isTrue(method.getAnnotation(Output.class) != null,
-									"A declarative @StreamListener method must be annotated with @Output");
-						}
-						invokeSetupMethodOnListenedChannel(method, bean);
+						registerHandlerMethodOnListenedChannel(method, streamListener, bean);
 					}
 				}
 			}
 		});
 		return bean;
 	}
 
+	private boolean isDeclarativeStreamListenerMethod(Method method, String methodAnnotatedInboundName, String methodAnnotatedOutboundName) {
+		int methodArgumentsLength = method.getParameterTypes().length;
+		for (int parameterIndex = 0; parameterIndex < methodArgumentsLength; parameterIndex++) {
+			MethodParameter methodParameter = MethodParameter.forMethodOrConstructor(method, parameterIndex);
+			if (methodParameter.hasParameterAnnotation(Input.class)) {
+				String inboundName = (String) AnnotationUtils.getValue(methodParameter.getParameterAnnotation(Input.class));
+				Assert.isTrue(StringUtils.hasText(inboundName), INVALID_INBOUND_NAME);
+				return isDeclarativeMethodParameter(getBindableBean(inboundName), methodParameter);
+			}
+			if (methodParameter.hasParameterAnnotation(Output.class)) {
+				String outboundName = (String) AnnotationUtils.getValue(methodParameter.getParameterAnnotation(Output.class));
+				Assert.isTrue(StringUtils.hasText(outboundName), INVALID_OUTBOUND_NAME);
+				return isDeclarativeMethodParameter(getBindableBean(outboundName), methodParameter);
+			}
+			if (StringUtils.hasText(methodAnnotatedOutboundName)) {
+				return isDeclarativeMethodParameter(getBindableBean(methodAnnotatedOutboundName), methodParameter);
+			}
+			if (StringUtils.hasText(methodAnnotatedInboundName)) {
+				return isDeclarativeMethodParameter(getBindableBean(methodAnnotatedInboundName), methodParameter);
+			}
+		}
+		return false;
+	}
+
+	private boolean isDeclarativeMethodParameter(Object targetBean, MethodParameter methodParameter) {
+		if (targetBean != null) {
+			if (methodParameter.getParameterType().isAssignableFrom(targetBean.getClass())) {
+				return true;
+			}
+			for (StreamListenerParameterAdapter<?, Object> streamListenerParameterAdapter : this.streamListenerParameterAdapters) {
+				if (streamListenerParameterAdapter.supports(targetBean.getClass(), methodParameter)) {
+					return true;
+				}
+			}
+		}
+		return false;
+	}
+
+	private Object getBindableBean(String boundElementName) {
```

Comment:
Actually, it is not clear for me why can't we just use `applicationContext.getBean()` instead and skip the local caching - the bean location process itself is very efficient.